## Visual Assessment: Acquaint Yourself
This Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`. Acquaint yourself with them through visual assessment below.

### Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

### Assess
In the cells below, each column of each table in this clinical trial dataset is described. To see the table that goes hand in hand with these descriptions, display each table in its entirety by displaying the pandas DataFrame that it was gathered into. This task is the mechanical part of visual assessment in pandas.

In [3]:
# Display the patients table
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Correia        2
Gersten        2
Ogochukwu      2
Collins        2
Tucker         2
Cindrić        2
Kadyrov        2
Silva          2
Grímsdóttir    2
Johnson        2
Nilsen         2
Lâm            2
Cabrera        2
Liễu           2
Batukayev      2
Souza          2
Dratchev       2
Lương          2
Aranda         2
Woźniak        2
Berg           2
Schiavone      2
Bùi            2
Hueber         2
Lund           2
Kowalczyk      2
Parker         2
              ..
Nakayama       1
Uspensky       1
Tikhonov       1
Villalpando    1
Chung          1
Grabljevec     1
Sung           1
Jørgensen      1
Reynoso        1
Pospíšil       1
Ekechukwu      1
Ono            1
Adams          1
Jackson        1
Tobeolisa      1
Adonay         1
Su             1
Harold         1
Holden         1
Wagner         1
Hsu            1
Lorenzo        1
Chin           1
Miranda        1
Klobučar       1
Horvat         1
Allaire        1
Dreher        

In [8]:
patients.address.value_counts()

123 Main Street                  6
2778 North Avenue                2
648 Old Dear Lane                2
2476 Fulton Street               2
1128 Jacobs Street               1
541 Green Hill Road              1
3414 Franklin Avenue             1
1854 Kyle Street                 1
2055 Emeral Dreams Drive         1
441 Tibbs Avenue                 1
4435 Poe Road                    1
2917 Golden Ridge Road           1
883 Oakwood Circle               1
3538 Paul Wayne Haggerty Road    1
4720 Gordon Street               1
4649 Worley Avenue               1
3450 Southern Street             1
1463 Beechwood Avenue            1
323 Platinum Drive               1
576 Brown Bear Drive             1
3857 Straford Park               1
1368 Yorkshire Circle            1
1975 Holden Street               1
1614 Heather Sees Way            1
1624 Rosebud Avenue              1
3662 Shinn Street                1
155 Red Hawk Road                1
2831 Milford Street              1
1619 Melm Street    

In [9]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [10]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [11]:
sum(treatments.auralin.isnull())

0

In [12]:
sum(treatments.novodra.isnull())

0

`patients` columns:
- **patient_id**: the unique identifier for each patient in the [Master Patient Index](https://en.wikipedia.org/wiki/Enterprise_master_patient_index) (i.e. patient database) of the pharmaceutical company that is producing Auralin
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year). The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

In [4]:
# Display the treatments table
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start` - `hba1c_end`. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [5]:
# Display the adverse_reactions table
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

`adverse_reactions` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **adverse_reaction**: the adverse reaction reported by the patient

Additional useful information:
- [Insulin resistance varies person to person](http://www.tudiabetes.org/forum/t/how-much-insulin-is-too-much-on-a-daily-basis/9804/5), which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This [diversity](https://www.clinicalleader.com/doc/an-fda-perspective-on-patient-diversity-in-clinical-trials-0001) is reflected in the `patients` table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a [similar debate](https://softwareengineering.stackexchange.com/questions/176582/is-there-an-excuse-for-short-variable-names) exists for variable names). The *auralin* and *novodra* column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

Quality:

- `patients` - Zipcode should be string.
- `patients` - Zipcode missing leading zeros
- `treatments`: missing HBA1C Changes values
- `patients` - Height '27' inch is entered incorrectly. 
- `patients` -  Some data has State Full Name and State Abb. Need to standardize the values for consistency.
- `patient` - Index 9 Patient name is misspelled as 'Dsvid'
- `patient` - Errorenous Data Types (Assigned Sex, State, Zip Code, birthdate)
- `pateints` - Contact Phone numbers are inconsistent.
- `patients` - John Doe have been entered 6 times. 
- `patients` - Multiple records for Jakobsen, Gersten, Taylor
- `patients` - Index 201 - has weight entered as KGs instead of pounds.
- `treatments` - 'Dash' data entry should be treated or entered as 'NaN'
- `treatments` - Has 'u' in the dose.
- `treatments` and `adverse reactions` - Lowercase names 
- `treatments` - has only 280 records.
- `treatments` - Errorenous Data Types (Auralin and Novodra Columns)
- `treatments` - Inaccurate HbA1C Change



Tidiness

- `patients` - Phone Number and Email should be separate columns.
- `treatments` - three columns in two columns. Treatment, Start Dose and End Dose.